staging/src/k8s.io/kubectl/pkg/cmd/profiling.go

```diff
@@ -32,60 +34,83 @@ var (
 )
 
 func addProfilingFlags(flags *pflag.FlagSet) {
-	flags.StringVar(&profileName, "profile", "none", "Name of profile to capture. One of (none|cpu|heap|goroutine|threadcreate|block|mutex)")
+	flags.StringVar(&profileName, "profile", "none", "Name of profile to capture. One of (none|cpu|heap|goroutine|threadcreate|block|mutex|trace)")
 	flags.StringVar(&profileOutput, "profile-output", "profile.pprof", "Name of the file to write the profile to")
 }
 
-func initProfiling() error {
+// initProfiling inits profiling and returns a function to be called on exit to flush and close.
+func initProfiling() (func() error, error) {
 	var (
 		f   *os.File
 		err error
 	)
 	switch profileName {
 	case "none":
-		return nil
+		return nil, nil
 	case "cpu":
 		f, err = os.Create(profileOutput)
 		if err != nil {
-			return err
+			return nil, err
 		}
+
 		err = pprof.StartCPUProfile(f)
 		if err != nil {
-			return err
+			f.Close() //nolint:errcheck
+			return nil, err
 		}
 	// Block and mutex profiles need a call to Set{Block,Mutex}ProfileRate to
 	// output anything. We choose to sample all events.
 	case "block":
 		runtime.SetBlockProfileRate(1)
 	case "mutex":
 		runtime.SetMutexProfileFraction(1)
+	case "trace":
+		f, err = os.Create(profileOutput)
+		if err != nil {
+			return nil, err
+		}
+
+		// Enable the CPU profiler. Samples will be captured in the execution trace.
+		// This is the same rate value as used in pprof.StartCPUProfile.
+		runtime.SetCPUProfileRate(100)
+		if err := trace.Start(f); err != nil {
+			f.Close() //nolint:errcheck
+			return nil, err
+		}
 	default:
 		// Check the profile name is valid.
 		if profile := pprof.Lookup(profileName); profile == nil {
-			return fmt.Errorf("unknown profile '%s'", profileName)
+			return nil, fmt.Errorf("unknown profile '%s'", profileName)
 		}
 	}
 
-	// If the command is interrupted before the end (ctrl-c), flush the
-	// profiling files
+	// If the command is interrupted before the end (ctrl-c), flush the profiling files
 	c := make(chan os.Signal, 1)
 	signal.Notify(c, os.Interrupt)
 	go func() {
 		<-c
-		f.Close()
-		flushProfiling()
+		flushProfiling(f) //nolint:errcheck
```

Comment:
I am not actually in love with this, if you are super unlucky, you can end up calling `flushProfiling` twice. What about wrapping in `sync.Once`?

Reply:
Yeah, there is a small probability. I'm not sure how wide this profiling is used. We can move forward as is, until someone complain about this.